src/middleware/jwtAuthentication.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+import jwt from 'jsonwebtoken'
+import logger from 'winston'
+
+import * as client from '../model/clients'
+import * as configIndex from '../config'
+import fs from 'fs'
+import path from 'path'
+
+const TOKEN_PATTERN = /^ *(?:[Bb][Ee][Aa][Rr][Ee][Rr]) +([A-Za-z0-9\-._~+/]+=*) *$/
+
+async function authenticateClient(clientID) {
+  return client.ClientModel.findOne({ clientID }).then((client) => {
+    if (!client) {
+      throw new Error('Client does not exist')
+    }
+    return client
+  })
+}
+
+function resolveJwtSecretOrPublicKey() {
+  let secretOrPublicKey = configIndex.config.get(
+    'authentication:jwt:secretOrPublicKey'
+  )
+
+  try {
+    const publicKeyFilePath = path.resolve(
+      __dirname,
+      '..',
+      '..',
+      'resources',
+      'certs',
+      'jwt',
+      secretOrPublicKey
+    )
+
+    // Check file exists
+    if (fs.existsSync(publicKeyFilePath)) {
+      secretOrPublicKey = fs.readFileSync(publicKeyFilePath).toString()
+    }
+    return secretOrPublicKey
+  } catch (error) {
+    throw new Error(
+      `Could not read public key file to verify asymmetric JWT: ${error}`
+    )
+  }
+}
+
+function getJwtOptions() {
+  const jwtConfig = configIndex.config.get('authentication:jwt')
+  const jwtOptions = {}
+
+  // Algorithms can be input as an environment variable therefore the string needs to be split
+  if (jwtConfig.algorithms) {
+    jwtOptions.algorithms = jwtConfig.algorithms.split(' ')
+  } else {
+    // The jsonwebtoken package does not require this field, but allowing any algorithm to be used opens a security risk
+    throw new Error('JWT Algorithm not specified')
+  }
+
+  // Audience can be input as an environment variable therefore the string needs to be split
+  if (jwtConfig.audience) {
+    jwtOptions.audience = jwtConfig.audience.split(' ')
+  }
+
+  jwtOptions.issuer = jwtConfig.issuer
+
+  return jwtOptions
+}
+
+async function authenticateToken(ctx) {
+  if (ctx.authenticated != null) {
+    return
+  }
+
+  const authHeader = ctx.request.header.authorization || ''
+  const token = TOKEN_PATTERN.exec(authHeader)
+
+  if (!token) {
+    logger.warn(`Missing or invalid 'Authorization' header`)
```

Comment:
Lets maybe specify which auth mechanism it is that failed incase all the auth mechanisms are enabled
`Missing or invalid 'JWT Authorization' header`